File: sdk/dotnet/Mics.HookSdk/MqEventDecoder.cs
using Google.Protobuf;
using Mics.Contracts.Hook.V1;
using Mics.Contracts.Message.V1;

namespace Mics.HookSdk;

public static class MqEventDecoder
{
    public static bool TryDecodeConnectAck(MqEvent evt, out ConnectAck ack)
    {
        ack = new ConnectAck();
        if (evt is null || evt.EventData.IsEmpty)
        {
            return false;
        }

        if (evt.EventType is not (EventType.ConnectOnline or EventType.ConnectOffline))
        {
            return false;
        }

        try
        {
            ack = ConnectAck.Parser.ParseFrom(evt.EventData);
            return true;
        }
        catch
        {
            return false;
        }
    }

    public static bool TryDecodeMessage(MqEvent evt, out MessageRequest message)
    {
        message = new MessageRequest();
        if (evt is null || evt.EventData.IsEmpty)
        {
            return false;
        }

        if (evt.EventType is not (EventType.SingleChatMsg or EventType.GroupChatMsg))
        {
            return false;
        }

        try
        {
            message = MessageRequest.Parser.ParseFrom(evt.EventData);
            return true;
        }
        catch
        {
            return false;
        }
    }

    public static bool TryVerifyAndDecodeConnectAck(string tenantSecret, MqEvent evt, out ConnectAck ack)
    {
        ack = new ConnectAck();
        if (!MqEventVerifier.Verify(tenantSecret, evt))
        {
            return false;
        }

        return TryDecodeConnectAck(evt, out ack);
    }

    public static bool TryVerifyAndDecodeMessage(string tenantSecret, MqEvent evt, out MessageRequest message)
    {
        message = new MessageRequest();
        if (!MqEventVerifier.Verify(tenantSecret, evt))
        {
            return false;
        }

        return TryDecodeMessage(evt, out message);
    }
}

File: sdk/java/mics-hook-sdk/src/main/java/com/mics/hooksdk/MqEventDecoder.java
package com.mics.hooksdk;

import com.google.protobuf.InvalidProtocolBufferException;
import com.mics.contracts.hook.v1.EventType;
import com.mics.contracts.hook.v1.MqEvent;
import com.mics.contracts.message.v1.ConnectAck;
import com.mics.contracts.message.v1.MessageRequest;

import java.util.Optional;

public final class MqEventDecoder {
    private MqEventDecoder() {
    }

    public static Optional<ConnectAck> tryDecodeConnectAck(MqEvent evt) {
        if (evt == null || evt.getEventData().isEmpty()) {
            return Optional.empty();
        }
        if (evt.getEventType() != EventType.CONNECT_ONLINE && evt.getEventType() != EventType.CONNECT_OFFLINE) {
            return Optional.empty();
        }
        try {
            return Optional.of(ConnectAck.parseFrom(evt.getEventData()));
        } catch (InvalidProtocolBufferException e) {
            return Optional.empty();
        }
    }

    public static Optional<MessageRequest> tryDecodeMessage(MqEvent evt) {
        if (evt == null || evt.getEventData().isEmpty()) {
            return Optional.empty();
        }
        if (evt.getEventType() != EventType.SINGLE_CHAT_MSG && evt.getEventType() != EventType.GROUP_CHAT_MSG) {
            return Optional.empty();
        }
        try {
            return Optional.of(MessageRequest.parseFrom(evt.getEventData()));
        } catch (InvalidProtocolBufferException e) {
            return Optional.empty();
        }
    }

    public static Optional<ConnectAck> tryVerifyAndDecodeConnectAck(String tenantSecret, MqEvent evt, boolean requireSign) {
        if (!MqEventSigner.verify(tenantSecret, evt, requireSign)) {
            return Optional.empty();
        }
        return tryDecodeConnectAck(evt);
    }

    public static Optional<MessageRequest> tryVerifyAndDecodeMessage(String tenantSecret, MqEvent evt, boolean requireSign) {
        if (!MqEventSigner.verify(tenantSecret, evt, requireSign)) {
            return Optional.empty();
        }
        return tryDecodeMessage(evt);
    }
}

